### ISSUE TITLE
Detect automatic dereferencing in pattern matches

### ISSUE THREAD

**Issue by kvark** [rust-lang/rust-clippy#6885]
### What it does

[RFC 2005](https://rust-lang.github.io/rfcs/2005-match-ergonomics.html) introduced the logic of auto-dereferencing into pattern matches. We find this harmful to both experts and beginners, and we avoid this extra logic in all project of gfx-rs ecosystem (gfx-rs, wgpu, wgpu-rs, naga, etc). It would help us to have a lint that we could enable in these projects.

Basically, the lint should fire whenever the pattern matching code *requires* the rules in RFC 2005 to compile. For example, on this code:
```rust
fn foo(e: &Foo) {
  match e {
    Foo {...}
  }
}
```

#### Justification

Code is read many more times than it's written. It's important to have the self-describing code, and being explicit in pattern matches helps that. For example, when I see `Foo(a, b)` in a pattern match, I want to know *exactly* what the type of `a` and `b` is. They can be what `type Foo` describes, or they can be references to these sub-types. With RFC 2005, it depends on whether or not the automatic dereferencing took place, which makes the program more confusing to read.

The original idea behind RFC 2005 was to allow people to express their intent with less code. What I see in practice, however, is that people new to the language end up writing *more* code, in addition to being confused about how the whole pattern matching works in Rust. I consider this feature harmful. Here is an example we had in Naga:
```rust
            let inner = match ty.as_ref() {
                TypeInner::Vector { kind, width, .. } | TypeInner::Matrix { kind, width, .. } => {
                    Cow::Owned(TypeInner::Scalar {
                        kind: *kind,
                        width: *width,
                    })
                }
```
This is a small piece, the problem spans on [multiple pages](https://github.com/gfx-rs/naga/blob/f7c5e4873e39f958a87d317e2c6a87a5f61e9d02/src/back/glsl.rs#L586) of code and multiple matches. People match on a reference, and then have to use `*` dereferences on all fields through the code. It makes me very sad to see, and it slows down their learning process of Rust.

### Categories (optional)

- Kind: `clippy::nursery`, `clippy::pedantic`

The recommended code would be easier to read and understand, since the identifies in a pattern match would correspond exactly to the types. In practice, it would be easier to write as well, since it would have less dereferences in the match arm bodies.

### Drawbacks

The lint would be at `Allow` level by default, so I don't think there are any drawbacks.

### Example

```rust
fn foo(e: &Foo) {
  match e {
    Foo(a, b) => {...}
  }
}
```

Could be written as:
```rust
fn foo(e: &Foo) {
  match *e {
    Foo(a, ref b) => {...}
  }
}
```

**Comment by nox** [comment]
@kvark I think this does what you want: https://rust-lang.github.io/rust-clippy/master/#pattern_type_mismatch

**Comment by kvark** [comment]
@nox beautiful, thank you!
I scanned through them and missed, expected something about dereferencing and automation in the name.
Looking forward to enable this lint everywhere :)
